### PR TITLE
Create user API to supply featurizer and model

### PR DESCRIPTION
**Pull request by gitconnoisseur**
This code allows users to get an atom coloration map by supplying a featurizer and trained model.

### CODE REVIEW

File: tests/test_create_map.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import os
+from rdkit import Chem
+from deepchem import deepchem as dc
+from xymol import XYMOL
+
+
+SMILES = "CCC1(CCC(C)C)C(=O)NC(=O)NC1=O"
+
+metric = dc.metrics.mean_absolute_error
+tasks, datasets, transformers = dc.molnet.load_delaney(featurizer='GraphConv', reload=False)
```
Comment (htian97) [inline tests/test_create_map.py:16]: Probably we do not have to use deepchem datasets. We can just create a dummy featurizer class/function to take any SMILES and return random vectors, through `featurize` API. Also, the ML model can also be a fake function to mimic real ones, take any vectors and return random prediction, through `predict` API.

File: setup.py [hunk 2]
```diff
@@ -37,6 +37,8 @@ def _get_version():
         "rdkit",
         "numpy",
         "matplotlib",
+        "deepchem",
```
Comment (htian97) [inline setup.py:40]: These are heavy dependencies that probably cause the CI failure.

File: xymol/xymol.py [hunk 3]
```diff
@@ -133,3 +135,43 @@ def plot_similarity_map(
...
         )
         img.savefig(file_name, bbox_inches='tight', dpi=600)
+
+    def create_map(
+        self, featurizer, model, filename="similarity_map.png"
+        ) -> list:
+        """Create similarity map
+
+        Args:
+            featurizer (class | str): if of type class, featurizer must have a featurize() function.
```
Comment (htian97) [inline xymol/xymol.py:145]: Please check other functions to have same comment style.

File: xymol/xymol.py [hunk 4]
```diff
@@ -133,3 +135,43 @@ def plot_similarity_map(
...
+            model (class | str): if of type class, model must have a predict() function.
+
+        Returns:
+            list of weights
+        """
+
+        if isinstance(featurizer, str):
```
Comment (htian97) [inline xymol/xymol.py:152]: We can just check if the input featurizer class/function has `featurize` API. It's the user's job to have this API.

Comment (htian97) [reply xymol/xymol.py:152]: And given in this case, we do not have to provide deepchem featurizers.

**Review by ghost — COMMENTED** [review]
Great start! Please address all comments.

File: xymol/xymol.py [hunk 5]
```diff
@@ -133,3 +134,35 @@ def plot_similarity_map(
...
+
+        Args:
+            featurizer (object): must have a featurize() function.
+            model (object): must have a predict() function.
+
+        Returns:
+            list: a list of weights
+        """
+        if not( hasattr(featurizer, "featurize") and callable(getattr(featurizer, "featurize")) ):
+            raise ValueError("Featurizer must have featurize() function.")
```
Comment (htian97) [inline xymol/xymol.py:151]: There should be at least one test on the wrong model and featurizer that do not have the corresponding APIs.

File: tests/test_create_map.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        if data[0][0] == 1:
+            return [0.0]
+
+        # return the test weight if predicting for drop_atoms
+        return [TEST_WEIGHT] * len(data)
+
+class TestFeaturizer:
+    def featurize(self, smiles) -> list:
+
+        #return list of zeros if drop_atoms is passed
```
Comment (htian97) [inline tests/test_create_map.py:28]: we should have a consistent comment style: one space after #. same below

File: xymol/xymol.py [hunk 7]
```diff
@@ -133,3 +134,35 @@ def plot_similarity_map(
...
+
+        drop_atoms = np.array(self.drop_each_atom())
+        drop_atoms_feat = np.array(featurizer.featurize(drop_atoms))
+
+        predictions = model.predict(drop_atoms_feat)
+
+        weights = [pred - parent_prediction for pred in predictions]
+        self.plot_similarity_map(weights, filename)
+
+        return weights
```
Comment (htian97) [inline xymol/xymol.py:168]: we probably do not need to return weights. This API is simply used for plotting.

**Review by ghost — COMMENTED** [review]
almost done, but there is still something we can improve. I will also check codecov to set a threshold.

File: xymol/xymol.py [hunk 8]
```diff
@@ -133,3 +134,35 @@ def plot_similarity_map(
...
+        """Create similarity map
+
+        Args:
+            featurizer (object): must have a featurize() function.
+            model (object): must have a predict() function.
+
+        Returns:
+            list: a list of weights
+        """
+        if not( hasattr(featurizer, "featurize") and callable(getattr(featurizer, "featurize")) ):
```
[the remark this answers is not in the document]
Comment (gitconnoisseur) [reply xymol/xymol.py:150]: do you mean the warning about the line being too long or is it an issue with the docstring?

File: xymol/xymol.py (hunk 7 above)
Comment (gitconnoisseur) [reply xymol/xymol.py:168]: OK, should I also remove these lines in the test code?
```   
    # check that all values in weights equal correct test difference (drop_atom pred minus parent_pred)
    assert weights[0] == TEST_WEIGHT
    assert weights.count(weights[0]) == len(weights)
```

File: xymol/xymol.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (htian97) [reply xymol/xymol.py:150]: line too long. We are following PEP8, max line length is 79.

File: xymol/xymol.py (hunk 7 above)
Comment (htian97) [reply xymol/xymol.py:168]: yes. Just try to check that there is a new png file.

**Review by ghost — APPROVED** [review]
Well done!

**Pull request merged**